File: dotnet/cs/src/Modulation/Cache.cs
/*
 * File: Cache.cs
 * Project: Modulation
 * Created Date: 13/09/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 07/12/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */

#if UNITY_2018_3_OR_NEWER
#define USE_SINGLE
#endif

using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Diagnostics.CodeAnalysis;
using AUTD3Sharp.NativeMethods;

namespace AUTD3Sharp.Modulation
{
    /// <summary>
    /// Modulation to cache the result of calculation
    /// </summary>
    public class Cache : Internal.Modulation, IEnumerable<EmitIntensity>, IDisposable
    {
        private bool _isDisposed;

        private CachePtr _cache;
        private readonly EmitIntensity[] _buffer;

        public Cache(Internal.Modulation m)
        {
            unsafe
            {
                _cache = NativeMethodsBase.AUTDModulationWithCache(m.ModulationPtr()).Validate();
                _buffer = new EmitIntensity[NativeMethodsBase.AUTDModulationCacheGetBufferLen(_cache)];
                fixed (EmitIntensity* p = &_buffer[0])
                    NativeMethodsBase.AUTDModulationCacheGetBuffer(_cache, (byte*)p);
            }
        }


        [ExcludeFromCodeCoverage]
        ~Cache()
        {
            Dispose();
        }

        public void Dispose()
        {
            if (_isDisposed) return;

            if (_cache.Item1 != IntPtr.Zero) NativeMethodsBase.AUTDModulationCacheDelete(_cache);
            _cache.Item1 = IntPtr.Zero;

            _isDisposed = true;
            GC.SuppressFinalize(this);
        }

        internal override ModulationPtr ModulationPtr()
        {
            return NativeMethodsBase.AUTDModulationCacheIntoModulation(_cache);
        }

        public EmitIntensity this[int index] => _buffer[index];

        public ReadOnlyCollection<EmitIntensity> Buffer => Array.AsReadOnly(_buffer);

        public IEnumerator<EmitIntensity> GetEnumerator()
        {
            foreach (var e in _buffer)
                yield return e;
        }


        [ExcludeFromCodeCoverage] System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => GetEnumerator();
    }

    public static class CacheModulationExtensions
    {
        public static Cache WithCache(this Internal.Modulation s)
        {
            return new Cache(s);
        }
    }
}

File: python/pyautd3/modulation/cache.py
"""
File: cache.py
Project: modulation
Created Date: 10/10/2023
Author: Shun Suzuki
-----
Last Modified: 10/10/2023
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2023 Shun Suzuki. All rights reserved.

"""


from collections.abc import Iterator
from typing import TypeVar

import numpy as np

from pyautd3.internal.modulation import IModulation
from pyautd3.internal.utils import _validate_ptr
from pyautd3.native_methods.autd3capi import NativeMethods as Base
from pyautd3.native_methods.autd3capi_def import CachePtr, ModulationPtr

M = TypeVar("M", bound=IModulation)


class Cache(IModulation):
    """Modulation to cache the result of calculation."""

    _cache: CachePtr
    _buffer: np.ndarray

    def __init__(self: "Cache", m: M) -> None:
        self._cache = _validate_ptr(Base().modulation_with_cache(m._modulation_ptr()))
        n = int(Base().modulation_cache_get_buffer_len(self._cache))
        self._buffer = np.zeros(n, dtype=np.uint8)
        Base().modulation_cache_get_buffer(self._cache, np.ctypeslib.as_ctypes(self._buffer))

    @property
    def buffer(self: "Cache") -> np.ndarray:
        """Get cached modulation data."""
        return self._buffer

    def __getitem__(self: "Cache", key: int) -> float:
        return self._buffer[key]

    def __iter__(self: "Cache") -> Iterator[float]:
        return iter(self._buffer)

    def _modulation_ptr(self: "Cache") -> ModulationPtr:
        return Base().modulation_cache_into_modulation(self._cache)

    def __del__(self: "Cache") -> None:
        Base().modulation_cache_delete(self._cache)


def __with_cache(self: M) -> Cache:
    """Cache the result of calculation."""
    return Cache(self)


IModulation.with_cache = __with_cache  # type: ignore[method-assign]
